File: typescript/session2/td/heritageMultiple/AlgebreNatRecursif.ts
import { Nat } from "../../../session1/td/naturels"
import { AlgebreNatRecursifSuccesseur } from "./AlgebreNatRecursifSuccesseur";
import { AlgebreNatRecursifZero } from "./AlgebreNatRecursifZero";

export interface AlgebreNatRecursif extends AlgebreNatRecursifSuccesseur, AlgebreNatRecursifZero {
    somme(val : Nat) : Nat;
    zero() : Nat;
    produit(x : Nat) : Nat;
    un() : Nat;
    modulo(x : Nat) : Nat
    div(x: Nat) : Nat;
    estEgal(x : Nat) : boolean;
    representation() : string;
}

export abstract class AlgebreNatRecursif_defaut implements AlgebreNatRecursif {
    abstract creerNatAvecValeur(valeur: number) : Nat;
    abstract creerNatAvecRepresentation(repDecimale : String) : Nat;
    abstract val() : number;
    abstract estNul() : boolean;
    abstract predecesseur() : Nat;
    abstract taille() : number;
    abstract chiffre(i : number) : number;
    abstract creerZero() : Nat;
    abstract creerSuccesseur(predecesseur : Nat) : Nat;
    
    somme(x : Nat) : Nat{
		if(this.estNul()) { 
			return x;
		}
		return this.creerSuccesseur(this.predecesseur().somme(x));
	}
	
	zero() : Nat{
		return this.creerZero();
	}
	
	produit(x : Nat) : Nat {
		if(this.estNul()) { 
			return this.creerZero();
		}
		return x.somme(this.predecesseur().produit(x));
	}
	
	un() : Nat {
		return this.creerNatAvecValeur(1);
	}
	
	modulo(x : Nat) : Nat{
		if(this.estNul()) { 
			return this.creerZero();
		}
		let r = this.predecesseur().modulo(x);
		return this.creerSuccesseur(r).estEgal(x) ? this.creerZero() : this.creerSuccesseur(r);
	}
	
	div(x : Nat) : Nat{
		if(this.estNul()) { 
			return this.creerZero();
		}
		let r = this.predecesseur().modulo(x);
		let q = this.predecesseur().div(x);
		return this.creerSuccesseur(r).estEgal(x) ? this.creerSuccesseur(q) : q;
	}
	
	estEgal(x : Nat) : boolean {
		return this.val() === x.val();
    }
    
    representation() : string {
        return String(this.val());
    }
}

File: java/cours/src/session2/td/heritageMultiple/AlgebreNatRecursif.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public interface AlgebreNatRecursif extends AlgebreNatRecursifSuccesseur, AlgebreNatRecursifZero {

	default public Nat somme(Nat val) {
		if(this.estNul()) { 
			return val;
		}
		return this.creerSuccesseur(this.predecesseur().somme(val));
	}
	
	default public Nat zero() {
		return this.creerZero();
	}
	
	default public Nat produit(Nat val) {
		if(this.estNul()) { 
			return this.creerZero();
		}
		return val.somme(this.predecesseur().produit(val));
	}
	
	default Nat un() {
		return this.creerNatAvecValeur(1);
	}
	
	default public Nat modulo(Nat val) {
		if(this.estNul()) { 
			return this.creerZero();
		}
		Nat r = this.predecesseur().modulo(val);
		return this.creerSuccesseur(r).equals(val) ? this.creerZero() : this.creerSuccesseur(r);
	}
	
	default public Nat div(Nat val) {
		if(this.estNul()) { 
			return this.creerZero();
		}
		Nat r = this.predecesseur().modulo(val);
		Nat q = this.predecesseur().div(val);
		return this.creerSuccesseur(r).equals(val) ? this.creerSuccesseur(q) : q;
	}
	
	default public boolean estEgal(Object o) {
		if(!(o instanceof Nat))
			return false;
		Nat x = (Nat)o;
		if(x.estNul())
			return false;
		return this.predecesseur().equals(x.predecesseur());
	} 	
}
